turbinia/workers/analysis/auth.py

```diff
@@ -0,0 +1,619 @@
+
+    # Generate result_markdown
+    markdown = []
+    markdown.append('## Brute Force Analysis\n')
+
+    for report in reports:
+      markdown.append(f'### Brute Force from {report["source_ip"]}\n')
+
+      for brute_force_login in report['brute_force_logins']:
+        markdown.append(
+            f'- Successful brute force from {brute_force_login["source_ip"]} as'
+            f' {brute_force_login["username"]} at'
+            f' {self.human_timestamp(brute_force_login["login_timestamp"])}'
+            f' (duration={brute_force_login["session_duration"]})')
+
+        if brute_force_login['session_duration'] > 600:
+          markdown.append(
+              '**NOTE**: Long login duration (>10 minutes).'
+              ' Potentially human activity')
+
+      markdown.append('\n#### IP Summaries\n')
+      for ip_summary in report['ip_summaries']:
+        markdown.append(f'- Source IP: {ip_summary["source_ip"]}')
+        markdown.append(
+            f'- Brute forcing IP first seen:'
+            f' {self.human_timestamp(ip_summary["first_seen"])}')
+        markdown.append(
+            f'- Brute forcing IP last seen:'
+            f' {self.human_timestamp(ip_summary["last_seen"])}')
+        markdown.append('- First successful login for brute forcing IP')
+        markdown.append(f'    - IP: {ip_summary["first_auth_ip"]}')
+        markdown.append(
+            f'    - Login timestamp:'
+            f' {self.human_timestamp(ip_summary["first_auth_timestamp"])}')
+        markdown.append(f'    - Username: {ip_summary["first_auth_username"]}')
+        markdown.append(
+            f'- Total successful login from IP:'
+            f' {ip_summary["total_success_events"]}')
+        markdown.append(
+            f'- Total failed login attempts:'
+            f' {ip_summary["total_failed_events"]}')
+
+        success_ip = ', '.join(ip_summary['success_source_ip_list'])
+        markdown.append(
+            f'- IP addresses that successfully logged in: {success_ip}')
+
+        success_usernames = ', '.join(ip_summary['success_username_list'])
+        markdown.append(
+            f'- Usernames that successfully logged in: {success_usernames}')
+        markdown.append(
+            f'- Total number of unique username attempted:'
+            f' {ip_summary["distinct_username_count"]}')
+        markdown.append('- Top 10 username attempted')
+        for username, count in ip_summary['top_usernames'].items():
+          markdown.append(f'    - {username}: {count}')
+
+    markdown.append('')
+    analyzer_output['result_markdown'] = '\n'.join(markdown)
+
+    return analyzer_output
+
+  def run(self, df: pd.DataFrame) -> dict:
+    """Entry point for the analyzer.
+
+    Args:
+      df (pd.DataFrame): Panda dataframe with authentication data
+
+    Returns:
+      AnalyzerResult: Result as AnalyzerResult object.
+    """
+    if df.empty:
+      raise Exception('[{self.NAME}] Dataframe is empty')
```

Comment:
Similar to the other exception raising, can we either raise a `TurbiniaException`, or if we want to raise something different, maybe `ValueError`?

Reply:
Probably want to add the exception to the docstring too.

Reply:
Fixed. Used TurbiniaException and added docstring.